pkg/apis/networking/types.go

```diff
@@ -56,13 +58,40 @@ type GroupMemberPod struct {
 	Ports []NamedPort
 }
 
+// ExternalEntityReference represents a ExternalEntity Reference.
+type ExternalEntityReference struct {
+	// The name of this ExternalEntity.
+	Name string
+	// The namespace of this ExternalEntity.
+	Namespace string
+}
+
+// ExternalEndpoint represents an external endpoint.
+type ExternalEndpoint struct {
+	// IP is the IP address of the ExternalEndpoint.
+	IP IPAddress
+	// Ports is the list NamedPort of the ExternalEndpoint.
+	Ports []NamedPort
+}
+
+// GroupMemberExternalEntity represents an ExternalEntity related member to be populated in Groups.
+type GroupMemberExternalEntity struct {
```

Comment:
@tnqn: again should we use an unified struct for both Pod and ExternalEntity?

Reply:
If they can share almost same fields and we can come up with a good name for the struct, I think we can have a unified struct.

Reply:
Again first do you think shared structs can enable more code sharing?

For GroupMemberExternalEntity and GroupMemberPod, one difference is that GroupMemberExternalEntity could include multiple endpoints, but GroupMemberPod includes only a single IP. Would you change Pod to use multiple endpoints as well?

Reply:
I am ok with code sharing, but I think implementation wise it may be too wide spread a change. Do u think it is reasonable that we define a new common structure; and this new common structure shall only be used for ExternalEntity passing for now. 

In subsequent releases, we can fully migrate to this new common structure?

Reply:
we can define something like

type NetworkEntityReference struct {
    name string
   namespace string
   kind string
}
type Endpoint struct {
	IP IPAddress
	Ports []NamedPort
}
type NetworkEntity struct {
   Reference NetworkEntityReference
   Endpoints []Endpoint
}
this can eventually be used to carry Pod info as well. But to limit the impact, we will only used it externalEntity, will that work?

Reply:
Yes, I understand if we change existing structs we will touch more existing code, but not certain if it is still better than adding a new set of code for the new structs. Would hear @tnqn's opinions.

Reply:
I got a question here:
From https://github.com/vmware-tanzu/antrea/pull/774/files#r433553833, `ExternalEntity` can have multiple IPs, but their NamedPorts are the same? If so, why we need to put `[]NamedPort` into `ExternalEndpoint` or `Endpoint` here?

If we are clear that `ExternalEntity` will have a single slice of NamedPorts for all IPs, then its struct can be simplified and be exactly the same as `GroupMemberPod` as a Pod could have multiple IPs when supporting dual-stack network. Renaming `GroupMemberPod` to `GroupMember` and changing `IP` to `IPs` can work for both of them. 
We can follow the K8s [Volume](https://github.com/kubernetes/kubernetes/blob/f2e3154a140dc192760a5cf8c01f8044e1aa867b/staging/src/k8s.io/api/core/v1/types.go#L37-L159) and [EnvVar](https://github.com/kubernetes/kubernetes/blob/f2e3154a140dc192760a5cf8c01f8044e1aa867b/staging/src/k8s.io/api/core/v1/types.go#L1837-L1874) structs, which also has multiple mutually exclusive source for the data.

```
type GroupMember struct {
	// Pod maintains the reference to the Pod.
	Pod *PodReference
	// ExternalEntity maintains the reference to the ExternalEntity.
	ExternalEntity *ExternalEntityReference
	// IP maintains the IPAddress of the member.
	IPs []IPAddress
	// Ports maintain the list of named port associated with this member.
	Ports []NamedPort
}
```
Or
```
type GroupMember struct {
        GroupMemberSource `json:",inline"`
	// IP maintains the IPAddress of the member.
	IPs []IPAddress
	// Ports maintain the list of named port associated with this member.
	Ports []NamedPort
}
type GroupMemberSource struct {
	// Pod maintains the reference to the Pod.
	Pod *PodReference
	// ExternalEntity maintains the reference to the ExternalEntity.
	ExternalEntity *ExternalEntityReference
}
```
With this struct, I think some of agent side code can focus on `IPs` and `Ports` fields and ignore their GroupMemberSource, this may address the requirement @suwang48404 is asking in the slack:
> do u mind I change the basic type CompletedRule.FromAddress/ToAddress and ruleCache.addressSetByGroup from "GroupMemberPod" to an new struct including only  [IP, port]? That way, the same framework may be extended to externalEntity too. It seems safe as agent does not use PodReference in GroupMemberPod if they are specified in To/From fields of the rule?

Reply:
1. I think the internal ExternalEntity defined here is  consistent with external facing one at https://github.com/vmware-tanzu/antrea/pull/716. I assume it is agreed upon. The reason that Abishek defined as such I assume it is because it is more flexible allowing different ports being exported on different IP?

2. whole we definitely can do as suggested to use the common GroupMemberSource . May I suggest we keep the existing PodGroupMember, so that we don't need to change a lot of agent/controller code?

Reply:
1. I apologize that I didn't notice this detail when reviewing #716, but I think it's unnecessary to make named port various for a single Entity. Originally we got NamedPort from all ports of all container of a Pod, and it's guaranteed to not conflict. Besides, named port is not used to describe which ports are exported but used to resolve port names for Policies. It doesn't require an ExternalEntity must have a named port when we have a Policy selecting it and allowing a port number, e.g. 80. A named port is only required when the Policy allows a port name, e.g. api.
2. If we agree on the common struct, I agree that you could probably just use the common struct for ExternalEntity and keep GroupMemberPod as is, we can then have another PR to migrate GroupMemberPod to the common struct.

Reply:
Regarding to 1), Initially I have the similar feeling, but I kind of grow to the current ExternalEndPoint as I believe @abhiraut must have thought it through. It allows us describes a use case where VM has multiple IPs, and some services(ports) exposed via one IP, others with another IP. Unlikely yes; possible, why not?

Reply:
IMO in this case the object would be a bit complicated to use. Say a VM have multiple IPs and user wants to define a named port "etcd" to "2379", the named port needs to be repeated in every Endpoint, otherwise agent won't resolve "etcd" to "2379" for those Endpoints that don't have this named port, while it would be redundant for most cases that a name maps to a port for all IPs. I think it's not very worth to ask redundant input for for most cases to support some unlikely needed cases.
I don't mean to make the change in this PR, but want to better understand the motivation of NamedPorts per IP as the API and code could be simpler without it. Any idea? @jianjuns @Dyanngg

Reply:
@Quan, What u said is true. 

I don't know the reason @abhiraut have, but I kind of like it because it offers flexibility. In most use cases, it is no issue because Pod/VM generally has single IP. In case there are multiple IPs on VM, for security reasons, it is common practice that services generally open only on primary IP.

Incidentally, since u r online, can u please approve this. I lost the approval while re push to deal with CI stability issue